_posts/2015-07-17-cassandra-write-performance-on-gce-and-aws.md

```diff
@@ -5,4 +5,91 @@ layout: post
 comments: true
 ---
 
-... or how we learned to create cross-cloud repeatable benchmarks for everybody. 
+*... or how we created cross-cloud tunable, repeatable benchmarks for everybody.*
+
+tl;dr - We achieved better Cassandra performance on GCE vs. Amazon, at close to half the cost.
+
+Benchmarking and performance are interesting problems, especially in today’s growing [microservices](https://en.wikipedia.org/wiki/Microservices) scene. It used to be a question of “how does this hardware compare to that hardware,” but as computing and service-oriented architectures grow the question has evolved. How does my [cloud and application](http://12factor.net/) stack handle this? It’s no longer enough to run your favorite machine benchmark tool on your web server and call it a day.
+
+Measuring every microservice in your stack, from backend to frontend, is a complex task. We started thinking about how you would model a system to benchmark all of these services. It’s not just a matter of measuring the performance of one service, but also its interactions with other services. Now multiply that by every config option for every service, like Cassandra, which has hundreds of options that can affect performance. And let’s not forget everything that you need to get Cassandra running, all those things have knobs too.
+
+So, to start this process, we decided to benchmark Cassandra on different public clouds. Over the next few months we’ll continue to expand for larger sizes and eventually we plan on rerunning the famous [1 million writes in Cassandra](http://googlecloudplatform.blogspot.com/2014/03/cassandra-hits-one-million-writes-per-second-on-google-compute-engine.html) across some public clouds. Let’s get started!
+
+#The Setup
+
+Every iteration of benchmark used the following constants:
+
+- Apache Cassandra 2.1 with OpenJDK 7 and Oracle Java SE 8u51
+- Ubuntu 14.04.1 with Juju 1.24
+
+We used the following hardware configuration on AWS:
+
+- 8 cores with 30G RAM (m3.2xlarge)
+
+and comparable hardware on GCE:
+
+- 8 cores with 30G RAM (n1-standard-8)
+
+## Juju
+Leveraging Juju to model the infrastructure, we're able to condense the installation, tuning, and benchmarking of any service into a composable, repeatable component to execute against any architecture or substrate. This allowed us to turn this:
+
+```
+cassandra-stress write n=2000000 cl=LOCAL_ONE -mode native cql3 -schema keyspace=Keyspace1 -log -node 52.3.185.174
```

Comment:
Yes, but this doesn't even touch on the deployment of cassandra, I'd include 

* Create cloud instances
* Read the [cassandra manual]
* Set up Cassandra
* [Tune Cassandra]
* Set up cassandra-stress
* stress code line

Into this:

* juju deploy cassandra -n3
* juju deploy cassandra-stress
* juju add-relation cassandra-stress cassandra
* juju action line